### PR TITLE
[ROCm] Fix distributed tests failure and enable ROCm distributed CI

### PR DESCRIPTION
**Pull request by jaglinux**
Distributed tests fails due to  AttributeError: 'torch._C._distributed_c10d.ProcessGroup'
object has no attribute '_set_backend' , when running distributed/test_c10d_spawn_gloo.py
This leads to tests not progressing resulting in hang.
Use _register_backend instead of _set_backend.

Fixes https://github.com/pytorch/pytorch/pull/91632

More details of issue: https://github.com/pytorch/pytorch/pull/91632#issuecomment-1402831950 and https://github.com/pytorch/pytorch/pull/91632#issuecomment-1405646977

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport

### CODE REVIEW

**Comment by jaglinux** [comment]
cc @jeffdaily @jithunnair-amd

**Comment by jaglinux** [comment]
The distributed test job on ROCm ran successfully.
https://github.com/pytorch/pytorch/actions/runs/4002638127/jobs/6870189994
2023-01-25T04:12:06.4216319Z Attribute _set_backend not found
2023-01-25T04:12:06.4216839Z Attribute _set_backend not found
The tests ran successfully after catching the exception.

**Review by jithunnair-amd — APPROVED** [review]
LGTM

File: test/distributed/test_c10d_spawn_gloo.py [hunk 1]
```diff
@@ -35,8 +35,11 @@ def _init_pg_gloo(cls, rank, filename, world_size):
             # set process group backends manually
             c10d.init_process_group(backend="gloo", store=store, rank=rank, world_size=world_size)
             pg = c10d.distributed_c10d._get_default_group()
-            pg._set_backend(torch.device("cpu"), c10d.Backend.GLOO, backend)
-            pg._set_backend(torch.device("cuda"), c10d.Backend.GLOO, backend)
+            try:
+                pg._set_backend(torch.device("cpu"), c10d.Backend.GLOO, backend)
+                pg._set_backend(torch.device("cuda"), c10d.Backend.GLOO, backend)
+            except AttributeError:
+                print('Attribute _set_backend not found')
```
Comment (malfet) [inline test/distributed/test_c10d_spawn_gloo.py:42]: c

**Review by malfet — CHANGES_REQUESTED** [review]
This change affects `test_c10d_spawn_gloo.py` behavior on all platforms, not just on rocm. Please add special case to the `except` blog that it will re-raise the exception if it is not running on ROCM.

Or better still, if those tests are no different for gloo backend, then for other, just skip them?

**Comment by jaglinux** [comment]
> This change affects `test_c10d_spawn_gloo.py` behavior on all platforms, not just on rocm. Please add special case to the `except` blog that it will re-raise the exception if it is not running on ROCM.
> 
> Or better still, if those tests are no different for gloo backend, then for other, just skip them?

_set_backend() is not defined in c10d
https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/init.cpp
These tests are not run with Python3.10 and hence not seen on other platforms. ROCm distributed tests run with Python3.8

More info about setting_backend, specifically for GLOO in [PR 88330](https://github.com/pytorch/pytorch/pull/88330/files#diff-136ce0cd6cca5b640f13f4d88cf9bde2001cb55ae69dffde6c7860b75b0fb604R38)

File: test/distributed/test_c10d_spawn_gloo.py (hunk 1 above)
Comment (jithunnair-amd) [reply test/distributed/test_c10d_spawn_gloo.py:42]: @H-Huang It seems you had added the `_set_backend` lines in your PR https://github.com/pytorch/pytorch/pull/88330. However, IIUC, those lines are only being tested on ROCm due to the python < 3.9 check, and they give an error `object has no attribute '_set_backend'`, which is not a ROCm-specific error afaict. Can you please suggest if there's a better fix that doesn't involve ROCm-specific logic?

Comment (H-Huang) [inline test/distributed/test_c10d_spawn_gloo.py:42]:
```suggestion
            pg._register_backend(torch.device("cpu"), c10d.ProcessGroup.BackendType.GLOO, backend)
            pg._register_backend(torch.device("cuda"), c10d.ProcessGroup.BackendType.GLOO, backend)
```

Comment (jaglinux) [reply test/distributed/test_c10d_spawn_gloo.py]: Thanks, pushed the commit.

**Review by H-Huang — APPROVED** [review]
Distributed test changes LGTM

**Comment by jaglinux** [comment]
The distributed tests (concerning this PR) has passed.
2023-01-31T22:11:01.7755496Z test_shared_allgather_chunk_gloo (__main__.ProcessGroupShareTensorTest)
2023-01-31T22:11:01.7757034Z test_shared_allgather_gloo (__main__.ProcessGroupShareTensorTest)
2023-01-31T22:11:01.7759039Z test_shared_allreduce_gloo (__main__.ProcessGroupShareTensorTest)
2023-01-31T22:11:01.7760815Z test_shared_broadcast_gloo (__main__.ProcessGroupShareTensorTest)

[Source](https://pipelines.actions.githubusercontent.com/serviceHosts/7d146c05-69c3-4c20-a0e7-818111670117/_apis/pipelines/1/runs/2598632/signedlogcontent/203?urlExpires=2023-02-01T01%3A31%3A49.8352445Z&urlSigningMethod=HMACV1&urlSignature=ilFgj4G8eckyaZ9gaeSXD%2FMyQFKrg2B8u5TY1vL8%2BX8%3D)

**Comment by jaglinux** [comment]
Not able to repro the issue locally. Retriggering the CI again for more data.

**Review by malfet — APPROVED** [review]

**Pull request closed without merging**